vllm/worker/model_runner_base.py

```diff
@@ -0,0 +1,205 @@
+        Override for fields that require some custom deserialization.
+        """
+        init_kwargs = {}
+        for field in dataclasses.fields(cls):
+            val = kwargs.get(field.name, None)
+            if val is not None:
+                init_kwargs[field.name] = val
+        return init_kwargs
+
+    @classmethod
+    def new(cls: Type[T], **kwargs) -> T:
+        """
+        Create a new instance of this class. Populate the new instance with
+        the given kwargs.
+        """
+        kwargs = cls._get_init_kwargs(**kwargs)
+        return cls(**kwargs)
+
+    def replace(self: T, **kwargs) -> T:
+        """
+        Replace current fields with fields in kwargs.
+        """
+        valid_kwargs = self.__class__._get_init_kwargs(**kwargs)
+        return dataclasses.replace(self, **valid_kwargs)
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        """
+        Extract broadcastable fields. Override for fields that require some
+        custom deserialization.
+        """
+        tensor_dict: Dict[str, Union[int, torch.Tensor]] = {}
+        for field in self.broadcastable_fields:
+            val = getattr(self, field, None)
+            if val is not None:
+                tensor_dict[field] = val
+
+        return tensor_dict
+
+    @staticmethod
+    def _add_attn_metadata_broadcastable_dict(
+            tensor_dict: Dict[str, Union[int, torch.Tensor]],
+            attn_metadata: Optional["AttentionMetadata"]) -> None:
+        """
+        Helper method to update tensor_dict with broadcastable
+        AttentionMetadata fields.
+        """
+        if attn_metadata is not None:
+            tensor_dict.update(attn_metadata.asdict_zerocopy())
+
+    @staticmethod
+    def _init_attn_metadata_from_kwargs(
+            attn_backend: Optional["AttentionBackend"] = None,
+            attn_metadata: Optional["AttentionMetadata"] = None,
+            **kwargs) -> Dict[str, Any]:
+        """
+        Helper method to initialize AttentionMetadata based on an
+        AttentionBackend and broadcastable AttentionMetadata fields.
+        """
+        if attn_metadata is None and attn_backend is not None:
+            # Extract the fields used to create AttentionMetadata.
+            valid_attn_kwargs = {}
+            for field in dataclasses.fields(attn_backend.get_metadata_cls()):
+                val = kwargs.pop(field.name, None)
+                if val is not None:
+                    valid_attn_kwargs[field.name] = val
+
+            attn_metadata = attn_backend.make_metadata(**valid_attn_kwargs)
+        if attn_metadata is not None:
+            kwargs["attn_metadata"] = attn_metadata
+        return kwargs
+
+    @staticmethod
+    def _init_sampling_metadata_from_kwargs(  # type: ignore
+            selected_token_indices: Optional[torch.Tensor] = None,
+            sampling_metadata: Optional["SamplingMetadata"] = None,
```

Comment:
why sampling metadata and attention metadata leak here?

Reply:
Moved these to helper methods.